// @aws-sdk/client-chime-sdk-messaging/dist-cjs/ChimeSDKMessaging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChimeSDKMessaging = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const ChimeSDKMessagingClient_1 = require("./ChimeSDKMessagingClient");
const AssociateChannelFlowCommand_1 = require("./commands/AssociateChannelFlowCommand");
const BatchCreateChannelMembershipCommand_1 = require("./commands/BatchCreateChannelMembershipCommand");
const ChannelFlowCallbackCommand_1 = require("./commands/ChannelFlowCallbackCommand");
const CreateChannelBanCommand_1 = require("./commands/CreateChannelBanCommand");
const CreateChannelCommand_1 = require("./commands/CreateChannelCommand");
const CreateChannelFlowCommand_1 = require("./commands/CreateChannelFlowCommand");
const CreateChannelMembershipCommand_1 = require("./commands/CreateChannelMembershipCommand");
const CreateChannelModeratorCommand_1 = require("./commands/CreateChannelModeratorCommand");
const DeleteChannelBanCommand_1 = require("./commands/DeleteChannelBanCommand");
const DeleteChannelCommand_1 = require("./commands/DeleteChannelCommand");
const DeleteChannelFlowCommand_1 = require("./commands/DeleteChannelFlowCommand");
const DeleteChannelMembershipCommand_1 = require("./commands/DeleteChannelMembershipCommand");
const DeleteChannelMessageCommand_1 = require("./commands/DeleteChannelMessageCommand");
const DeleteChannelModeratorCommand_1 = require("./commands/DeleteChannelModeratorCommand");
const DeleteMessagingStreamingConfigurationsCommand_1 = require("./commands/DeleteMessagingStreamingConfigurationsCommand");
const DescribeChannelBanCommand_1 = require("./commands/DescribeChannelBanCommand");
const DescribeChannelCommand_1 = require("./commands/DescribeChannelCommand");
const DescribeChannelFlowCommand_1 = require("./commands/DescribeChannelFlowCommand");
const DescribeChannelMembershipCommand_1 = require("./commands/DescribeChannelMembershipCommand");
const DescribeChannelMembershipForAppInstanceUserCommand_1 = require("./commands/DescribeChannelMembershipForAppInstanceUserCommand");
const DescribeChannelModeratedByAppInstanceUserCommand_1 = require("./commands/DescribeChannelModeratedByAppInstanceUserCommand");
const DescribeChannelModeratorCommand_1 = require("./commands/DescribeChannelModeratorCommand");
const DisassociateChannelFlowCommand_1 = require("./commands/DisassociateChannelFlowCommand");
const GetChannelMembershipPreferencesCommand_1 = require("./commands/GetChannelMembershipPreferencesCommand");
const GetChannelMessageCommand_1 = require("./commands/GetChannelMessageCommand");
const GetChannelMessageStatusCommand_1 = require("./commands/GetChannelMessageStatusCommand");
const GetMessagingSessionEndpointCommand_1 = require("./commands/GetMessagingSessionEndpointCommand");
const GetMessagingStreamingConfigurationsCommand_1 = require("./commands/GetMessagingStreamingConfigurationsCommand");
const ListChannelBansCommand_1 = require("./commands/ListChannelBansCommand");
const ListChannelFlowsCommand_1 = require("./commands/ListChannelFlowsCommand");
const ListChannelMembershipsCommand_1 = require("./commands/ListChannelMembershipsCommand");
const ListChannelMembershipsForAppInstanceUserCommand_1 = require("./commands/ListChannelMembershipsForAppInstanceUserCommand");
const ListChannelMessagesCommand_1 = require("./commands/ListChannelMessagesCommand");
const ListChannelModeratorsCommand_1 = require("./commands/ListChannelModeratorsCommand");
const ListChannelsAssociatedWithChannelFlowCommand_1 = require("./commands/ListChannelsAssociatedWithChannelFlowCommand");
const ListChannelsCommand_1 = require("./commands/ListChannelsCommand");
const ListChannelsModeratedByAppInstanceUserCommand_1 = require("./commands/ListChannelsModeratedByAppInstanceUserCommand");
const ListSubChannelsCommand_1 = require("./commands/ListSubChannelsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const PutChannelExpirationSettingsCommand_1 = require("./commands/PutChannelExpirationSettingsCommand");
const PutChannelMembershipPreferencesCommand_1 = require("./commands/PutChannelMembershipPreferencesCommand");
const PutMessagingStreamingConfigurationsCommand_1 = require("./commands/PutMessagingStreamingConfigurationsCommand");
const RedactChannelMessageCommand_1 = require("./commands/RedactChannelMessageCommand");
const SearchChannelsCommand_1 = require("./commands/SearchChannelsCommand");
const SendChannelMessageCommand_1 = require("./commands/SendChannelMessageCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateChannelCommand_1 = require("./commands/UpdateChannelCommand");
const UpdateChannelFlowCommand_1 = require("./commands/UpdateChannelFlowCommand");
const UpdateChannelMessageCommand_1 = require("./commands/UpdateChannelMessageCommand");
const UpdateChannelReadMarkerCommand_1 = require("./commands/UpdateChannelReadMarkerCommand");
const commands = {
    AssociateChannelFlowCommand: AssociateChannelFlowCommand_1.AssociateChannelFlowCommand,
    BatchCreateChannelMembershipCommand: BatchCreateChannelMembershipCommand_1.BatchCreateChannelMembershipCommand,
    ChannelFlowCallbackCommand: ChannelFlowCallbackCommand_1.ChannelFlowCallbackCommand,
    CreateChannelCommand: CreateChannelCommand_1.CreateChannelCommand,
    CreateChannelBanCommand: CreateChannelBanCommand_1.CreateChannelBanCommand,
    CreateChannelFlowCommand: CreateChannelFlowCommand_1.CreateChannelFlowCommand,
    CreateChannelMembershipCommand: CreateChannelMembershipCommand_1.CreateChannelMembershipCommand,
    CreateChannelModeratorCommand: CreateChannelModeratorCommand_1.CreateChannelModeratorCommand,
    DeleteChannelCommand: DeleteChannelCommand_1.DeleteChannelCommand,
    DeleteChannelBanCommand: DeleteChannelBanCommand_1.DeleteChannelBanCommand,
    DeleteChannelFlowCommand: DeleteChannelFlowCommand_1.DeleteChannelFlowCommand,
    DeleteChannelMembershipCommand: DeleteChannelMembershipCommand_1.DeleteChannelMembershipCommand,
    DeleteChannelMessageCommand: DeleteChannelMessageCommand_1.DeleteChannelMessageCommand,
    DeleteChannelModeratorCommand: DeleteChannelModeratorCommand_1.DeleteChannelModeratorCommand,
    DeleteMessagingStreamingConfigurationsCommand: DeleteMessagingStreamingConfigurationsCommand_1.DeleteMessagingStreamingConfigurationsCommand,
    DescribeChannelCommand: DescribeChannelCommand_1.DescribeChannelCommand,
    DescribeChannelBanCommand: DescribeChannelBanCommand_1.DescribeChannelBanCommand,
    DescribeChannelFlowCommand: DescribeChannelFlowCommand_1.DescribeChannelFlowCommand,
    DescribeChannelMembershipCommand: DescribeChannelMembershipCommand_1.DescribeChannelMembershipCommand,
    DescribeChannelMembershipForAppInstanceUserCommand: DescribeChannelMembershipForAppInstanceUserCommand_1.DescribeChannelMembershipForAppInstanceUserCommand,
    DescribeChannelModeratedByAppInstanceUserCommand: DescribeChannelModeratedByAppInstanceUserCommand_1.DescribeChannelModeratedByAppInstanceUserCommand,
    DescribeChannelModeratorCommand: DescribeChannelModeratorCommand_1.DescribeChannelModeratorCommand,
    DisassociateChannelFlowCommand: DisassociateChannelFlowCommand_1.DisassociateChannelFlowCommand,
    GetChannelMembershipPreferencesCommand: GetChannelMembershipPreferencesCommand_1.GetChannelMembershipPreferencesCommand,
    GetChannelMessageCommand: GetChannelMessageCommand_1.GetChannelMessageCommand,
    GetChannelMessageStatusCommand: GetChannelMessageStatusCommand_1.GetChannelMessageStatusCommand,
    GetMessagingSessionEndpointCommand: GetMessagingSessionEndpointCommand_1.GetMessagingSessionEndpointCommand,
    GetMessagingStreamingConfigurationsCommand: GetMessagingStreamingConfigurationsCommand_1.GetMessagingStreamingConfigurationsCommand,
    ListChannelBansCommand: ListChannelBansCommand_1.ListChannelBansCommand,
    ListChannelFlowsCommand: ListChannelFlowsCommand_1.ListChannelFlowsCommand,
    ListChannelMembershipsCommand: ListChannelMembershipsCommand_1.ListChannelMembershipsCommand,
    ListChannelMembershipsForAppInstanceUserCommand: ListChannelMembershipsForAppInstanceUserCommand_1.ListChannelMembershipsForAppInstanceUserCommand,
    ListChannelMessagesCommand: ListChannelMessagesCommand_1.ListChannelMessagesCommand,
    ListChannelModeratorsCommand: ListChannelModeratorsCommand_1.ListChannelModeratorsCommand,
    ListChannelsCommand: ListChannelsCommand_1.ListChannelsCommand,
    ListChannelsAssociatedWithChannelFlowCommand: ListChannelsAssociatedWithChannelFlowCommand_1.ListChannelsAssociatedWithChannelFlowCommand,
    ListChannelsModeratedByAppInstanceUserCommand: ListChannelsModeratedByAppInstanceUserCommand_1.ListChannelsModeratedByAppInstanceUserCommand,
    ListSubChannelsCommand: ListSubChannelsCommand_1.ListSubChannelsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    PutChannelExpirationSettingsCommand: PutChannelExpirationSettingsCommand_1.PutChannelExpirationSettingsCommand,
    PutChannelMembershipPreferencesCommand: PutChannelMembershipPreferencesCommand_1.PutChannelMembershipPreferencesCommand,
    PutMessagingStreamingConfigurationsCommand: PutMessagingStreamingConfigurationsCommand_1.PutMessagingStreamingConfigurationsCommand,
    RedactChannelMessageCommand: RedactChannelMessageCommand_1.RedactChannelMessageCommand,
    SearchChannelsCommand: SearchChannelsCommand_1.SearchChannelsCommand,
    SendChannelMessageCommand: SendChannelMessageCommand_1.SendChannelMessageCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateChannelCommand: UpdateChannelCommand_1.UpdateChannelCommand,
    UpdateChannelFlowCommand: UpdateChannelFlowCommand_1.UpdateChannelFlowCommand,
    UpdateChannelMessageCommand: UpdateChannelMessageCommand_1.UpdateChannelMessageCommand,
    UpdateChannelReadMarkerCommand: UpdateChannelReadMarkerCommand_1.UpdateChannelReadMarkerCommand,
};
class ChimeSDKMessaging extends ChimeSDKMessagingClient_1.ChimeSDKMessagingClient {
}
exports.ChimeSDKMessaging = ChimeSDKMessaging;
(0, smithy_client_1.createAggregatedClient)(commands, ChimeSDKMessaging);
