// @aws-sdk/client-chime-sdk-messaging/dist-cjs/models/models_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelModeratorSummaryFilterSensitiveLog = exports.ChannelModeratorFilterSensitiveLog = exports.ChannelModeratedByAppInstanceUserSummaryFilterSensitiveLog = exports.ChannelMessageSummaryFilterSensitiveLog = exports.ChannelMessageFilterSensitiveLog = exports.ChannelMembershipSummaryFilterSensitiveLog = exports.ChannelMembershipPreferencesFilterSensitiveLog = exports.PushNotificationPreferencesFilterSensitiveLog = exports.ChannelMembershipForAppInstanceUserSummaryFilterSensitiveLog = exports.ChannelSummaryFilterSensitiveLog = exports.ChannelMembershipFilterSensitiveLog = exports.ChannelFlowSummaryFilterSensitiveLog = exports.ChannelFlowCallbackRequestFilterSensitiveLog = exports.ChannelMessageCallbackFilterSensitiveLog = exports.PushNotificationConfigurationFilterSensitiveLog = exports.MessageAttributeValueFilterSensitiveLog = exports.ChannelFlowFilterSensitiveLog = exports.ProcessorFilterSensitiveLog = exports.ChannelBanSummaryFilterSensitiveLog = exports.ChannelBanFilterSensitiveLog = exports.ChannelAssociatedWithFlowSummaryFilterSensitiveLog = exports.ChannelFilterSensitiveLog = exports.BatchCreateChannelMembershipResponseFilterSensitiveLog = exports.BatchChannelMembershipsFilterSensitiveLog = exports.IdentityFilterSensitiveLog = exports.SearchFieldOperator = exports.SearchFieldKey = exports.SortOrder = exports.MessagingDataType = exports.ChannelMessageType = exports.ChannelMessageStatus = exports.ChannelMessagePersistenceType = exports.PushNotificationType = exports.FallbackAction = exports.InvocationType = exports.ChannelPrivacy = exports.ChannelMode = exports.ExpirationCriterion = exports.ResourceLimitExceededException = exports.UnauthorizedClientException = exports.ThrottledClientException = exports.ServiceUnavailableException = exports.ServiceFailureException = exports.NotFoundException = exports.ForbiddenException = exports.ConflictException = exports.BadRequestException = exports.ErrorCode = exports.ChannelMembershipType = exports.AllowNotifications = void 0;
exports.UpdateChannelMessageRequestFilterSensitiveLog = exports.UpdateChannelFlowRequestFilterSensitiveLog = exports.UpdateChannelRequestFilterSensitiveLog = exports.UntagResourceRequestFilterSensitiveLog = exports.TagResourceRequestFilterSensitiveLog = exports.SendChannelMessageRequestFilterSensitiveLog = exports.SearchChannelsResponseFilterSensitiveLog = exports.SearchChannelsRequestFilterSensitiveLog = exports.PutChannelMembershipPreferencesResponseFilterSensitiveLog = exports.PutChannelMembershipPreferencesRequestFilterSensitiveLog = exports.ListTagsForResourceResponseFilterSensitiveLog = exports.ListSubChannelsResponseFilterSensitiveLog = exports.ListSubChannelsRequestFilterSensitiveLog = exports.ListChannelsModeratedByAppInstanceUserResponseFilterSensitiveLog = exports.ListChannelsModeratedByAppInstanceUserRequestFilterSensitiveLog = exports.ListChannelsAssociatedWithChannelFlowResponseFilterSensitiveLog = exports.ListChannelsAssociatedWithChannelFlowRequestFilterSensitiveLog = exports.ListChannelsResponseFilterSensitiveLog = exports.ListChannelsRequestFilterSensitiveLog = exports.ListChannelModeratorsResponseFilterSensitiveLog = exports.ListChannelModeratorsRequestFilterSensitiveLog = exports.ListChannelMessagesResponseFilterSensitiveLog = exports.ListChannelMessagesRequestFilterSensitiveLog = exports.ListChannelMembershipsForAppInstanceUserResponseFilterSensitiveLog = exports.ListChannelMembershipsForAppInstanceUserRequestFilterSensitiveLog = exports.ListChannelMembershipsResponseFilterSensitiveLog = exports.ListChannelMembershipsRequestFilterSensitiveLog = exports.ListChannelFlowsResponseFilterSensitiveLog = exports.ListChannelFlowsRequestFilterSensitiveLog = exports.ListChannelBansResponseFilterSensitiveLog = exports.ListChannelBansRequestFilterSensitiveLog = exports.GetChannelMessageResponseFilterSensitiveLog = exports.GetChannelMembershipPreferencesResponseFilterSensitiveLog = exports.DescribeChannelModeratorResponseFilterSensitiveLog = exports.DescribeChannelModeratedByAppInstanceUserResponseFilterSensitiveLog = exports.DescribeChannelMembershipForAppInstanceUserResponseFilterSensitiveLog = exports.DescribeChannelMembershipResponseFilterSensitiveLog = exports.DescribeChannelFlowResponseFilterSensitiveLog = exports.DescribeChannelBanResponseFilterSensitiveLog = exports.DescribeChannelResponseFilterSensitiveLog = exports.CreateChannelModeratorResponseFilterSensitiveLog = exports.CreateChannelMembershipResponseFilterSensitiveLog = exports.CreateChannelFlowRequestFilterSensitiveLog = exports.CreateChannelBanResponseFilterSensitiveLog = exports.CreateChannelRequestFilterSensitiveLog = exports.TagFilterSensitiveLog = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const ChimeSDKMessagingServiceException_1 = require("./ChimeSDKMessagingServiceException");
exports.AllowNotifications = {
    ALL: "ALL",
    FILTERED: "FILTERED",
    NONE: "NONE",
};
exports.ChannelMembershipType = {
    DEFAULT: "DEFAULT",
    HIDDEN: "HIDDEN",
};
exports.ErrorCode = {
    AccessDenied: "AccessDenied",
    BadRequest: "BadRequest",
    Conflict: "Conflict",
    Forbidden: "Forbidden",
    NotFound: "NotFound",
    PhoneNumberAssociationsExist: "PhoneNumberAssociationsExist",
    PreconditionFailed: "PreconditionFailed",
    ResourceLimitExceeded: "ResourceLimitExceeded",
    ServiceFailure: "ServiceFailure",
    ServiceUnavailable: "ServiceUnavailable",
    Throttled: "Throttled",
    Throttling: "Throttling",
    Unauthorized: "Unauthorized",
    Unprocessable: "Unprocessable",
    VoiceConnectorGroupAssociationsExist: "VoiceConnectorGroupAssociationsExist",
};
class BadRequestException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "BadRequestException",
            $fault: "client",
            ...opts,
        });
        this.name = "BadRequestException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BadRequestException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.BadRequestException = BadRequestException;
class ConflictException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        this.name = "ConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConflictException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.ConflictException = ConflictException;
class ForbiddenException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "ForbiddenException",
            $fault: "client",
            ...opts,
        });
        this.name = "ForbiddenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ForbiddenException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.ForbiddenException = ForbiddenException;
class NotFoundException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "NotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "NotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NotFoundException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.NotFoundException = NotFoundException;
class ServiceFailureException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "ServiceFailureException",
            $fault: "server",
            ...opts,
        });
        this.name = "ServiceFailureException";
        this.$fault = "server";
        Object.setPrototypeOf(this, ServiceFailureException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.ServiceFailureException = ServiceFailureException;
class ServiceUnavailableException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "ServiceUnavailableException",
            $fault: "server",
            ...opts,
        });
        this.name = "ServiceUnavailableException";
        this.$fault = "server";
        Object.setPrototypeOf(this, ServiceUnavailableException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.ServiceUnavailableException = ServiceUnavailableException;
class ThrottledClientException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "ThrottledClientException",
            $fault: "client",
            ...opts,
        });
        this.name = "ThrottledClientException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ThrottledClientException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.ThrottledClientException = ThrottledClientException;
class UnauthorizedClientException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "UnauthorizedClientException",
            $fault: "client",
            ...opts,
        });
        this.name = "UnauthorizedClientException";
        this.$fault = "client";
        Object.setPrototypeOf(this, UnauthorizedClientException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.UnauthorizedClientException = UnauthorizedClientException;
class ResourceLimitExceededException extends ChimeSDKMessagingServiceException_1.ChimeSDKMessagingServiceException {
    constructor(opts) {
        super({
            name: "ResourceLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceLimitExceededException.prototype);
        this.Code = opts.Code;
        this.Message = opts.Message;
    }
}
exports.ResourceLimitExceededException = ResourceLimitExceededException;
exports.ExpirationCriterion = {
    CREATED_TIMESTAMP: "CREATED_TIMESTAMP",
    LAST_MESSAGE_TIMESTAMP: "LAST_MESSAGE_TIMESTAMP",
};
exports.ChannelMode = {
    RESTRICTED: "RESTRICTED",
    UNRESTRICTED: "UNRESTRICTED",
};
exports.ChannelPrivacy = {
    PRIVATE: "PRIVATE",
    PUBLIC: "PUBLIC",
};
exports.InvocationType = {
    ASYNC: "ASYNC",
};
exports.FallbackAction = {
    ABORT: "ABORT",
    CONTINUE: "CONTINUE",
};
exports.PushNotificationType = {
    DEFAULT: "DEFAULT",
    VOIP: "VOIP",
};
exports.ChannelMessagePersistenceType = {
    NON_PERSISTENT: "NON_PERSISTENT",
    PERSISTENT: "PERSISTENT",
};
exports.ChannelMessageStatus = {
    DENIED: "DENIED",
    FAILED: "FAILED",
    PENDING: "PENDING",
    SENT: "SENT",
};
exports.ChannelMessageType = {
    CONTROL: "CONTROL",
    STANDARD: "STANDARD",
};
exports.MessagingDataType = {
    Channel: "Channel",
    ChannelMessage: "ChannelMessage",
};
exports.SortOrder = {
    ASCENDING: "ASCENDING",
    DESCENDING: "DESCENDING",
};
exports.SearchFieldKey = {
    MEMBERS: "MEMBERS",
};
exports.SearchFieldOperator = {
    EQUALS: "EQUALS",
    INCLUDES: "INCLUDES",
};
const IdentityFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
});
exports.IdentityFilterSensitiveLog = IdentityFilterSensitiveLog;
const BatchChannelMembershipsFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.InvitedBy && { InvitedBy: (0, exports.IdentityFilterSensitiveLog)(obj.InvitedBy) }),
    ...(obj.Members && { Members: obj.Members.map((item) => (0, exports.IdentityFilterSensitiveLog)(item)) }),
});
exports.BatchChannelMembershipsFilterSensitiveLog = BatchChannelMembershipsFilterSensitiveLog;
const BatchCreateChannelMembershipResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.BatchChannelMemberships && {
        BatchChannelMemberships: (0, exports.BatchChannelMembershipsFilterSensitiveLog)(obj.BatchChannelMemberships),
    }),
});
exports.BatchCreateChannelMembershipResponseFilterSensitiveLog = BatchCreateChannelMembershipResponseFilterSensitiveLog;
const ChannelFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.CreatedBy && { CreatedBy: (0, exports.IdentityFilterSensitiveLog)(obj.CreatedBy) }),
});
exports.ChannelFilterSensitiveLog = ChannelFilterSensitiveLog;
const ChannelAssociatedWithFlowSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
});
exports.ChannelAssociatedWithFlowSummaryFilterSensitiveLog = ChannelAssociatedWithFlowSummaryFilterSensitiveLog;
const ChannelBanFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
    ...(obj.CreatedBy && { CreatedBy: (0, exports.IdentityFilterSensitiveLog)(obj.CreatedBy) }),
});
exports.ChannelBanFilterSensitiveLog = ChannelBanFilterSensitiveLog;
const ChannelBanSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
});
exports.ChannelBanSummaryFilterSensitiveLog = ChannelBanSummaryFilterSensitiveLog;
const ProcessorFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
});
exports.ProcessorFilterSensitiveLog = ProcessorFilterSensitiveLog;
const ChannelFlowFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Processors && { Processors: obj.Processors.map((item) => (0, exports.ProcessorFilterSensitiveLog)(item)) }),
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
});
exports.ChannelFlowFilterSensitiveLog = ChannelFlowFilterSensitiveLog;
const MessageAttributeValueFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.StringValues && { StringValues: smithy_client_1.SENSITIVE_STRING }),
});
exports.MessageAttributeValueFilterSensitiveLog = MessageAttributeValueFilterSensitiveLog;
const PushNotificationConfigurationFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Title && { Title: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Body && { Body: smithy_client_1.SENSITIVE_STRING }),
});
exports.PushNotificationConfigurationFilterSensitiveLog = PushNotificationConfigurationFilterSensitiveLog;
const ChannelMessageCallbackFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Content && { Content: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.PushNotification && {
        PushNotification: (0, exports.PushNotificationConfigurationFilterSensitiveLog)(obj.PushNotification),
    }),
    ...(obj.MessageAttributes && {
        MessageAttributes: Object.entries(obj.MessageAttributes).reduce((acc, [key, value]) => ((acc[key] = (0, exports.MessageAttributeValueFilterSensitiveLog)(value)), acc), {}),
    }),
    ...(obj.ContentType && { ContentType: smithy_client_1.SENSITIVE_STRING }),
});
exports.ChannelMessageCallbackFilterSensitiveLog = ChannelMessageCallbackFilterSensitiveLog;
const ChannelFlowCallbackRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelMessage && { ChannelMessage: (0, exports.ChannelMessageCallbackFilterSensitiveLog)(obj.ChannelMessage) }),
});
exports.ChannelFlowCallbackRequestFilterSensitiveLog = ChannelFlowCallbackRequestFilterSensitiveLog;
const ChannelFlowSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Processors && { Processors: obj.Processors.map((item) => (0, exports.ProcessorFilterSensitiveLog)(item)) }),
});
exports.ChannelFlowSummaryFilterSensitiveLog = ChannelFlowSummaryFilterSensitiveLog;
const ChannelMembershipFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.InvitedBy && { InvitedBy: (0, exports.IdentityFilterSensitiveLog)(obj.InvitedBy) }),
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
});
exports.ChannelMembershipFilterSensitiveLog = ChannelMembershipFilterSensitiveLog;
const ChannelSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
});
exports.ChannelSummaryFilterSensitiveLog = ChannelSummaryFilterSensitiveLog;
const ChannelMembershipForAppInstanceUserSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelSummary && { ChannelSummary: (0, exports.ChannelSummaryFilterSensitiveLog)(obj.ChannelSummary) }),
});
exports.ChannelMembershipForAppInstanceUserSummaryFilterSensitiveLog = ChannelMembershipForAppInstanceUserSummaryFilterSensitiveLog;
const PushNotificationPreferencesFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.FilterRule && { FilterRule: smithy_client_1.SENSITIVE_STRING }),
});
exports.PushNotificationPreferencesFilterSensitiveLog = PushNotificationPreferencesFilterSensitiveLog;
const ChannelMembershipPreferencesFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.PushNotifications && {
        PushNotifications: (0, exports.PushNotificationPreferencesFilterSensitiveLog)(obj.PushNotifications),
    }),
});
exports.ChannelMembershipPreferencesFilterSensitiveLog = ChannelMembershipPreferencesFilterSensitiveLog;
const ChannelMembershipSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
});
exports.ChannelMembershipSummaryFilterSensitiveLog = ChannelMembershipSummaryFilterSensitiveLog;
const ChannelMessageFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Content && { Content: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Sender && { Sender: (0, exports.IdentityFilterSensitiveLog)(obj.Sender) }),
    ...(obj.MessageAttributes && {
        MessageAttributes: Object.entries(obj.MessageAttributes).reduce((acc, [key, value]) => ((acc[key] = (0, exports.MessageAttributeValueFilterSensitiveLog)(value)), acc), {}),
    }),
    ...(obj.ContentType && { ContentType: smithy_client_1.SENSITIVE_STRING }),
});
exports.ChannelMessageFilterSensitiveLog = ChannelMessageFilterSensitiveLog;
const ChannelMessageSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Content && { Content: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Sender && { Sender: (0, exports.IdentityFilterSensitiveLog)(obj.Sender) }),
    ...(obj.MessageAttributes && {
        MessageAttributes: Object.entries(obj.MessageAttributes).reduce((acc, [key, value]) => ((acc[key] = (0, exports.MessageAttributeValueFilterSensitiveLog)(value)), acc), {}),
    }),
    ...(obj.ContentType && { ContentType: smithy_client_1.SENSITIVE_STRING }),
});
exports.ChannelMessageSummaryFilterSensitiveLog = ChannelMessageSummaryFilterSensitiveLog;
const ChannelModeratedByAppInstanceUserSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelSummary && { ChannelSummary: (0, exports.ChannelSummaryFilterSensitiveLog)(obj.ChannelSummary) }),
});
exports.ChannelModeratedByAppInstanceUserSummaryFilterSensitiveLog = ChannelModeratedByAppInstanceUserSummaryFilterSensitiveLog;
const ChannelModeratorFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Moderator && { Moderator: (0, exports.IdentityFilterSensitiveLog)(obj.Moderator) }),
    ...(obj.CreatedBy && { CreatedBy: (0, exports.IdentityFilterSensitiveLog)(obj.CreatedBy) }),
});
exports.ChannelModeratorFilterSensitiveLog = ChannelModeratorFilterSensitiveLog;
const ChannelModeratorSummaryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Moderator && { Moderator: (0, exports.IdentityFilterSensitiveLog)(obj.Moderator) }),
});
exports.ChannelModeratorSummaryFilterSensitiveLog = ChannelModeratorSummaryFilterSensitiveLog;
const TagFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Key && { Key: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Value && { Value: smithy_client_1.SENSITIVE_STRING }),
});
exports.TagFilterSensitiveLog = TagFilterSensitiveLog;
const CreateChannelRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.ClientRequestToken && { ClientRequestToken: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Tags && { Tags: obj.Tags.map((item) => (0, exports.TagFilterSensitiveLog)(item)) }),
    ...(obj.ChannelId && { ChannelId: smithy_client_1.SENSITIVE_STRING }),
});
exports.CreateChannelRequestFilterSensitiveLog = CreateChannelRequestFilterSensitiveLog;
const CreateChannelBanResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
});
exports.CreateChannelBanResponseFilterSensitiveLog = CreateChannelBanResponseFilterSensitiveLog;
const CreateChannelFlowRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Processors && { Processors: obj.Processors.map((item) => (0, exports.ProcessorFilterSensitiveLog)(item)) }),
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Tags && { Tags: obj.Tags.map((item) => (0, exports.TagFilterSensitiveLog)(item)) }),
    ...(obj.ClientRequestToken && { ClientRequestToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.CreateChannelFlowRequestFilterSensitiveLog = CreateChannelFlowRequestFilterSensitiveLog;
const CreateChannelMembershipResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
});
exports.CreateChannelMembershipResponseFilterSensitiveLog = CreateChannelMembershipResponseFilterSensitiveLog;
const CreateChannelModeratorResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelModerator && { ChannelModerator: (0, exports.IdentityFilterSensitiveLog)(obj.ChannelModerator) }),
});
exports.CreateChannelModeratorResponseFilterSensitiveLog = CreateChannelModeratorResponseFilterSensitiveLog;
const DescribeChannelResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Channel && { Channel: (0, exports.ChannelFilterSensitiveLog)(obj.Channel) }),
});
exports.DescribeChannelResponseFilterSensitiveLog = DescribeChannelResponseFilterSensitiveLog;
const DescribeChannelBanResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelBan && { ChannelBan: (0, exports.ChannelBanFilterSensitiveLog)(obj.ChannelBan) }),
});
exports.DescribeChannelBanResponseFilterSensitiveLog = DescribeChannelBanResponseFilterSensitiveLog;
const DescribeChannelFlowResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelFlow && { ChannelFlow: (0, exports.ChannelFlowFilterSensitiveLog)(obj.ChannelFlow) }),
});
exports.DescribeChannelFlowResponseFilterSensitiveLog = DescribeChannelFlowResponseFilterSensitiveLog;
const DescribeChannelMembershipResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelMembership && { ChannelMembership: (0, exports.ChannelMembershipFilterSensitiveLog)(obj.ChannelMembership) }),
});
exports.DescribeChannelMembershipResponseFilterSensitiveLog = DescribeChannelMembershipResponseFilterSensitiveLog;
const DescribeChannelMembershipForAppInstanceUserResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelMembership && {
        ChannelMembership: (0, exports.ChannelMembershipForAppInstanceUserSummaryFilterSensitiveLog)(obj.ChannelMembership),
    }),
});
exports.DescribeChannelMembershipForAppInstanceUserResponseFilterSensitiveLog = DescribeChannelMembershipForAppInstanceUserResponseFilterSensitiveLog;
const DescribeChannelModeratedByAppInstanceUserResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Channel && { Channel: (0, exports.ChannelModeratedByAppInstanceUserSummaryFilterSensitiveLog)(obj.Channel) }),
});
exports.DescribeChannelModeratedByAppInstanceUserResponseFilterSensitiveLog = DescribeChannelModeratedByAppInstanceUserResponseFilterSensitiveLog;
const DescribeChannelModeratorResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelModerator && { ChannelModerator: (0, exports.ChannelModeratorFilterSensitiveLog)(obj.ChannelModerator) }),
});
exports.DescribeChannelModeratorResponseFilterSensitiveLog = DescribeChannelModeratorResponseFilterSensitiveLog;
const GetChannelMembershipPreferencesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
    ...(obj.Preferences && { Preferences: (0, exports.ChannelMembershipPreferencesFilterSensitiveLog)(obj.Preferences) }),
});
exports.GetChannelMembershipPreferencesResponseFilterSensitiveLog = GetChannelMembershipPreferencesResponseFilterSensitiveLog;
const GetChannelMessageResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelMessage && { ChannelMessage: (0, exports.ChannelMessageFilterSensitiveLog)(obj.ChannelMessage) }),
});
exports.GetChannelMessageResponseFilterSensitiveLog = GetChannelMessageResponseFilterSensitiveLog;
const ListChannelBansRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelBansRequestFilterSensitiveLog = ListChannelBansRequestFilterSensitiveLog;
const ListChannelBansResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.ChannelBans && { ChannelBans: obj.ChannelBans.map((item) => (0, exports.ChannelBanSummaryFilterSensitiveLog)(item)) }),
});
exports.ListChannelBansResponseFilterSensitiveLog = ListChannelBansResponseFilterSensitiveLog;
const ListChannelFlowsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelFlowsRequestFilterSensitiveLog = ListChannelFlowsRequestFilterSensitiveLog;
const ListChannelFlowsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelFlows && { ChannelFlows: obj.ChannelFlows.map((item) => (0, exports.ChannelFlowSummaryFilterSensitiveLog)(item)) }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelFlowsResponseFilterSensitiveLog = ListChannelFlowsResponseFilterSensitiveLog;
const ListChannelMembershipsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelMembershipsRequestFilterSensitiveLog = ListChannelMembershipsRequestFilterSensitiveLog;
const ListChannelMembershipsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelMemberships && {
        ChannelMemberships: obj.ChannelMemberships.map((item) => (0, exports.ChannelMembershipSummaryFilterSensitiveLog)(item)),
    }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelMembershipsResponseFilterSensitiveLog = ListChannelMembershipsResponseFilterSensitiveLog;
const ListChannelMembershipsForAppInstanceUserRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelMembershipsForAppInstanceUserRequestFilterSensitiveLog = ListChannelMembershipsForAppInstanceUserRequestFilterSensitiveLog;
const ListChannelMembershipsForAppInstanceUserResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ChannelMemberships && {
        ChannelMemberships: obj.ChannelMemberships.map((item) => (0, exports.ChannelMembershipForAppInstanceUserSummaryFilterSensitiveLog)(item)),
    }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelMembershipsForAppInstanceUserResponseFilterSensitiveLog = ListChannelMembershipsForAppInstanceUserResponseFilterSensitiveLog;
const ListChannelMessagesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelMessagesRequestFilterSensitiveLog = ListChannelMessagesRequestFilterSensitiveLog;
const ListChannelMessagesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.ChannelMessages && {
        ChannelMessages: obj.ChannelMessages.map((item) => (0, exports.ChannelMessageSummaryFilterSensitiveLog)(item)),
    }),
});
exports.ListChannelMessagesResponseFilterSensitiveLog = ListChannelMessagesResponseFilterSensitiveLog;
const ListChannelModeratorsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelModeratorsRequestFilterSensitiveLog = ListChannelModeratorsRequestFilterSensitiveLog;
const ListChannelModeratorsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.ChannelModerators && {
        ChannelModerators: obj.ChannelModerators.map((item) => (0, exports.ChannelModeratorSummaryFilterSensitiveLog)(item)),
    }),
});
exports.ListChannelModeratorsResponseFilterSensitiveLog = ListChannelModeratorsResponseFilterSensitiveLog;
const ListChannelsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelsRequestFilterSensitiveLog = ListChannelsRequestFilterSensitiveLog;
const ListChannelsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Channels && { Channels: obj.Channels.map((item) => (0, exports.ChannelSummaryFilterSensitiveLog)(item)) }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelsResponseFilterSensitiveLog = ListChannelsResponseFilterSensitiveLog;
const ListChannelsAssociatedWithChannelFlowRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelsAssociatedWithChannelFlowRequestFilterSensitiveLog = ListChannelsAssociatedWithChannelFlowRequestFilterSensitiveLog;
const ListChannelsAssociatedWithChannelFlowResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Channels && {
        Channels: obj.Channels.map((item) => (0, exports.ChannelAssociatedWithFlowSummaryFilterSensitiveLog)(item)),
    }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelsAssociatedWithChannelFlowResponseFilterSensitiveLog = ListChannelsAssociatedWithChannelFlowResponseFilterSensitiveLog;
const ListChannelsModeratedByAppInstanceUserRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelsModeratedByAppInstanceUserRequestFilterSensitiveLog = ListChannelsModeratedByAppInstanceUserRequestFilterSensitiveLog;
const ListChannelsModeratedByAppInstanceUserResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Channels && {
        Channels: obj.Channels.map((item) => (0, exports.ChannelModeratedByAppInstanceUserSummaryFilterSensitiveLog)(item)),
    }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListChannelsModeratedByAppInstanceUserResponseFilterSensitiveLog = ListChannelsModeratedByAppInstanceUserResponseFilterSensitiveLog;
const ListSubChannelsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListSubChannelsRequestFilterSensitiveLog = ListSubChannelsRequestFilterSensitiveLog;
const ListSubChannelsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.ListSubChannelsResponseFilterSensitiveLog = ListSubChannelsResponseFilterSensitiveLog;
const ListTagsForResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Tags && { Tags: obj.Tags.map((item) => (0, exports.TagFilterSensitiveLog)(item)) }),
});
exports.ListTagsForResourceResponseFilterSensitiveLog = ListTagsForResourceResponseFilterSensitiveLog;
const PutChannelMembershipPreferencesRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Preferences && { Preferences: (0, exports.ChannelMembershipPreferencesFilterSensitiveLog)(obj.Preferences) }),
});
exports.PutChannelMembershipPreferencesRequestFilterSensitiveLog = PutChannelMembershipPreferencesRequestFilterSensitiveLog;
const PutChannelMembershipPreferencesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Member && { Member: (0, exports.IdentityFilterSensitiveLog)(obj.Member) }),
    ...(obj.Preferences && { Preferences: (0, exports.ChannelMembershipPreferencesFilterSensitiveLog)(obj.Preferences) }),
});
exports.PutChannelMembershipPreferencesResponseFilterSensitiveLog = PutChannelMembershipPreferencesResponseFilterSensitiveLog;
const SearchChannelsRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.SearchChannelsRequestFilterSensitiveLog = SearchChannelsRequestFilterSensitiveLog;
const SearchChannelsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Channels && { Channels: obj.Channels.map((item) => (0, exports.ChannelSummaryFilterSensitiveLog)(item)) }),
    ...(obj.NextToken && { NextToken: smithy_client_1.SENSITIVE_STRING }),
});
exports.SearchChannelsResponseFilterSensitiveLog = SearchChannelsResponseFilterSensitiveLog;
const SendChannelMessageRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Content && { Content: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.ClientRequestToken && { ClientRequestToken: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.PushNotification && {
        PushNotification: (0, exports.PushNotificationConfigurationFilterSensitiveLog)(obj.PushNotification),
    }),
    ...(obj.MessageAttributes && {
        MessageAttributes: Object.entries(obj.MessageAttributes).reduce((acc, [key, value]) => ((acc[key] = (0, exports.MessageAttributeValueFilterSensitiveLog)(value)), acc), {}),
    }),
    ...(obj.ContentType && { ContentType: smithy_client_1.SENSITIVE_STRING }),
});
exports.SendChannelMessageRequestFilterSensitiveLog = SendChannelMessageRequestFilterSensitiveLog;
const TagResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Tags && { Tags: obj.Tags.map((item) => (0, exports.TagFilterSensitiveLog)(item)) }),
});
exports.TagResourceRequestFilterSensitiveLog = TagResourceRequestFilterSensitiveLog;
const UntagResourceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.TagKeys && { TagKeys: smithy_client_1.SENSITIVE_STRING }),
});
exports.UntagResourceRequestFilterSensitiveLog = UntagResourceRequestFilterSensitiveLog;
const UpdateChannelRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
});
exports.UpdateChannelRequestFilterSensitiveLog = UpdateChannelRequestFilterSensitiveLog;
const UpdateChannelFlowRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Processors && { Processors: obj.Processors.map((item) => (0, exports.ProcessorFilterSensitiveLog)(item)) }),
    ...(obj.Name && { Name: smithy_client_1.SENSITIVE_STRING }),
});
exports.UpdateChannelFlowRequestFilterSensitiveLog = UpdateChannelFlowRequestFilterSensitiveLog;
const UpdateChannelMessageRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Content && { Content: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.Metadata && { Metadata: smithy_client_1.SENSITIVE_STRING }),
    ...(obj.ContentType && { ContentType: smithy_client_1.SENSITIVE_STRING }),
});
exports.UpdateChannelMessageRequestFilterSensitiveLog = UpdateChannelMessageRequestFilterSensitiveLog;
